### PR TITLE
[Feature] Hidden component and System

### PR DESCRIPTION
**Pull request by Telzhaak**
A simple `Hidden` component that allows hiding an entity,
and a `HideHierarchySystem` (based on `UiTransformSystem`)  that propagates the `Hidden` changes through all children.

### CODE REVIEW

**Comment by Telzhaak** [comment]
I have yet to verify whether the `(..., &visibility.visible_unordered, !&hidden).join()`'s are necessary, as the hidden entities shouldn't be part of `visibility.visible_ordered` nor `.visible_unordered`.

**Comment by jojolepro** [comment]
You can keep the profiler macro by the way, I'll need to add them absolutely everywhere soon :D

**Comment by Telzhaak** [comment]
Verified for passes
- DrawFlat
- DrawShaded
- DrawFlatSeperated
- DrawShadedSeperated
- DrawSprite
- DrawUi

Missing verification for
- DrawPbm
- DrawPbmSeperate
- DrawDebugLines

**Comment by Telzhaak** [comment]
Thanks to @Rhuagh the reason for crashing is fixed.

The dependency "parent_hierarchy_system" was missing, which is a hard necessity for the `HideHierarchySystem` to work. Therefore the system has been moved from utils to the renderer and integrated as an opt-in for the `RenderBundle`.

**Comment by jojolepro** [comment]
You can access all child entities of a specifix entity using specs-hierarchy. Having hidden qork on multiple frames will definitely create relly ugly looking glitches

**Comment by Telzhaak** [comment]
I did overlook those functions, it seems.

Now the propagation happens in a single run

File: amethyst_renderer/src/hide_system.rs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                let parent_entity = parents.get(*entity).unwrap().entity;
+                let parent_dirty = self.marked_as_modified.contains(parent_entity.id());
+                if parent_dirty {
+                    if hidden.contains(parent_entity) {
+                        for child in hierarchy.all_children_iter(parent_entity) {
+                            match hidden.insert(child, HiddenPropagate::default()) {
+                                Ok(_v) => (),
+                                Err(e) => {
+                                    error!(
+                                        "Entity: {:?}, {:?}",
```
Comment (Xaeroxe) [inline amethyst_renderer/src/hide_system.rs]: Can we make this a bit more descriptive of what operation we're trying to perform here?

File: amethyst_renderer/src/hide_system.rs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                        }
+                    }
+                }else if self_dirty {
+                    if hidden.contains(*entity) {
+                        for child in hierarchy.all_children_iter(*entity) {
+                            match hidden.insert(child, HiddenPropagate::default()) {
+                                Ok(_v) => (),
+                                Err(e) => {
+                                    error!(
+                                        "Entity: {:?}, {:?}",
```
Comment (Xaeroxe) [inline amethyst_renderer/src/hide_system.rs]: Same here

Comment (Telzhaak) [reply amethyst_renderer/src/hide_system.rs]: Okay, made it a little more verbose

File: amethyst_renderer/src/hide_system.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
+use amethyst_core::{
+    specs::prelude::{
+        BitSet, InsertedFlag, ReadExpect, ReadStorage, ReaderId, RemovedFlag, Resources, System,
+        WriteStorage,
+    },
+    transform::components::{HierarchyEvent, Parent, ParentHierarchy},
+};
+use HiddenPropagate;
+
+// Based on the [UiTransformSystem](struct.UiTransformSystem.html).
```
Comment (jojolepro) [inline amethyst_renderer/src/hide_system.rs]: Missing one /

File: amethyst_renderer/src/hide_system.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        for entity in hierarchy.all() {
+            {
+                let self_dirty = self.marked_as_modified.contains(entity.id());
+
+                let parent_entity = parents.get(*entity).unwrap().entity;
+                let parent_dirty = self.marked_as_modified.contains(parent_entity.id());
+                if parent_dirty {
+                    if hidden.contains(parent_entity) {
+                        for child in hierarchy.all_children_iter(parent_entity) {
+                            match hidden.insert(child, HiddenPropagate::default()) {
```
Comment (jojolepro) [inline amethyst_renderer/src/hide_system.rs]: blabla.map_err(|e| error!("msg {}", e))

File: amethyst_renderer/src/hide_system.rs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                                    );
+                                }
+                            }
+                        }
+                    } else {
+                        for child in hierarchy.all_children_iter(parent_entity) {
+                            hidden.remove(child);
+                        }
+                    }
+                }else if self_dirty {
```
Comment (jojolepro) [inline amethyst_renderer/src/hide_system.rs]: Someone forgot to format :P
There's a missing space after "}"

File: amethyst_renderer/src/hide_system.rs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                        for child in hierarchy.all_children_iter(parent_entity) {
+                            hidden.remove(child);
+                        }
+                    }
+                }else if self_dirty { 
+                    // in case the parent was already dirty, this entity and its children have already been hidden, 
+                    // therefore it only needs to be an else-if, instead of a stand-alone if.
+                    if hidden.contains(*entity) {
+                        for child in hierarchy.all_children_iter(*entity) {
+                            match hidden.insert(child, HiddenPropagate::default()) {
```
Comment (jojolepro) [inline amethyst_renderer/src/hide_system.rs]: .map_err

File: amethyst_renderer/src/hidden.rs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+pub struct Hidden;
+
+impl Component for Hidden {
+    type Storage = NullStorage<Self>;
+}
+
+/// Like [Hidden](struct.Hidden.html), but can propagate through children when the [HideHierarchySystem](struct.HideHierarchySystem.html)
+/// is enabled in the [RenderBundle](struct.RenderBundle.html).
+#[derive(Clone, Debug, Default)]
+pub struct HiddenPropagate;
```
Comment (jojolepro) [inline amethyst_renderer/src/hidden.rs]: Note: I do like the separation of having one that propagates and the other one who doesn't. :+1:

File: amethyst_ui/src/pass.rs [hunk 8]
```diff
@@ -144,13 +146,17 @@ impl Pass for DrawUi {
...
         ): <Self as PassData>::Data,
     ) {
         // Populate and update the draw order cache.
         {
             let bitset = &mut self.cached_draw_order.cached;
             self.cached_draw_order.cache.retain(|&(_z, entity)| {
-                let keep = ui_transform.contains(entity);
+                let keep = ui_transform.contains(entity)
+                    && !hidden.contains(entity)
+                    && !hidden_prop.contains(entity);
```
Comment (jojolepro) [inline amethyst_ui/src/pass.rs]: Did you test using the ui example to ensure you could hide and unhide with success?

File: amethyst_ui/src/pass.rs [hunk 9]
```diff
@@ -209,9 +222,9 @@ impl Pass for DrawUi {
         };
         effect.data.vertex_bufs.push(vbuf);
 
-        let highest_abs_z = (&ui_transform,)
+        let highest_abs_z = (&ui_transform, !&hidden)
```
Comment (jojolepro) [inline amethyst_ui/src/pass.rs]: missing hidden propagate?

File: amethyst_ui/src/pass.rs (hunk 8 above)
Comment (Telzhaak) [reply amethyst_ui/src/pass.rs]: Yes, I tested every example with `Hidden`, `HiddenPropagate` (with and without the `HideHierarchySystem` enabled).

Plus my own game's menu's, which are UiPrefabs

File: amethyst_renderer/src/hide_system.rs (hunk 4 above)
Comment (Telzhaak) [reply amethyst_renderer/src/hide_system.rs]: using `map_err` does not help, as it returns a `Result<T,E>` which has to be handled again.
The `?` operator is not implemented for `amethyst::error`

Comment (jojolepro) [reply amethyst_renderer/src/hide_system.rs]: Then you can
```rs
if let Err(e) = hierarchy.all_chi....{

}
```

**Comment by Telzhaak** [comment]
Addressed all change requests

- [x] cargo fmt
- [x] cargo build
- [x] cargo build --release
- [x] cargo test --all

**Comment by Telzhaak** [comment]
If noone except me tested it, then it's only verified on windows yet!

**Comment by jojolepro** [comment]
No reason to test other platforms if you don't touch platform dependent code

**Comment by jojolepro** [comment]
bors r+
:b:
might conflict after the other bors finishes, we'll see

**Pull request merged**